### ISSUE TITLE
[css-sizing] how should inline-axis intrinsic sizing work for 'fit-content' and 'fit-content()'?

### ISSUE THREAD

**Issue by dbaron** [w3c/csswg-drafts#3731]
This issue is probably blocking implementation of the `fit-content()` function (which is in css-sizing-3) in Gecko, based on [a bug comment](https://bugzilla.mozilla.org/show_bug.cgi?id=1312588#c9) I wrote a few months ago.  I realize I never filed an issue here about it, though, so I'm doing so now.

The css-sizing-4 spec's [definition of intrinsic sizes](https://drafts.csswg.org/css-sizing-4/#block-intrinsic) doesn't say how intrinsic sizing should treat the `fit-content()` function.  It also defines a behavior for the `fit-content` keyword that I'm not convinced makes sense.

What I wrote in the above bugzilla comment about how I think things ought to behave was the following:

> Intrinsic sizing behavior for this function (and for 'min-content', 'max-content', and 'fit-content' keywords) in the block dimension is not particularly relevant here, since they're all treated as the initial values.  while intrinsic sizing in the block dimension isn't particularly well defined until after https://github.com/w3c/csswg-drafts/issues/2890 happens, we don't need to worry about that here.
>
> I think the intrinsic sizing behavior for 'min-content' and 'max-content' in the inline dimension is pretty easy; they compute to a single length, and the intrinsic sizing behavior should be the same as for &lt;length> type values (for whichever of 'inline-size', 'min-inline-size', or 'max-inline-size' is the case -- &lt;length> values have different effects on the three of them).
>
> Intrinsic sizing for 'fit-content(&lt;length>)' is also easy, since it also computes to a single length.
>
> However, intrinsic sizing for 'fit-content(&lt;percentage>)' is harder, since what it resolves to in the end (during layout) depends on things *outside* the box.  All the previous cases were easy because the size specified is a function only of the contents of the element and the properties on it, but now we need to consider what to do for percentages.  Some of this may depend on how we treat inline-size: &lt;percentage>, min-inline-size: &lt;percentage>, and max-inline-size: &lt;percentage>.  However, there are also cases where the way we treat those was a compromise that we made for compatibility, and we might want to do something better here.  (On the flip side, doing something different might also be confusing if developers are accustomed to the current behavior.)  In the past, there have been three ways to approach handling &lt;percentage> in intrinsic widths:
>  (1) treat them just like 'auto' (i.e., pretend they're not there) -- this way has basically won (for percentages) at this point
>  (2) reverse-compute them on their own, where a box with max-content size S and width: P (a percentage, expressed so 1.0 is 100%) has a max-content contribution to its parent of S/P.  This prevents overlap in block-like layout, but isn't sufficient when there's more than one thing per line.
>  (3) reverse-compute them together, so that if a line/row/table-row has boxes with sizes S1...SN and percentages P1...PN and then other non-percentage (fixed) sized boxes F1...FN, the contribution of the entire row to its parent is the larger of sum(F1...FN)/(1 - max(sum(P1...PN), 1.0)) and any of Si/Pi.  The first term ensures that the sizes for the non-percentage parts fit within the percentage remaining after all the percentage parts, and the remaining terms ensure that each percentage part is big enough.  This is roughly what we do for tables.
>
> So (1) is simple for percentages -- but it's not immediately clear to me how it works for 'fit-content(&lt;percentage>)'.  That's the part that I'd need to think about more.  Doing (2) or (3) is probably even more complicated.  But there's probably a simpler way to think about what's needed for fit-content() that I haven't quite thought of yet...
>
> Intrinsic sizing for the 'fit-content' keyword is, I think, basically the same as 'fit-content(&lt;percentage>)', except that it has to deal with padding/border/margin differently.  But there's also a little less justification for it to behave like &lt;percentage> does today.

So, in other words, I think it's easy to specify the inline-axis intrinsic size behavior for `fit-content(<length>)` although the spec doesn't do so today.  It would just need to be added to the [clause](https://drafts.csswg.org/css-sizing-4/#block-intrinsic) that now says:

> If the computed inline-size of a block-level box is min-content, max-content, or a definite size ...

I think it's harder to specify the behavior for `fit-content(<percentage>)` (which the spec doesn't specify today) or for `fit-content` (which the spec currently says is the same as `auto` or `stretch`... but it's not clear to me that that's a good behavior).

**Comment by fantasai** [comment]
Sorry, I missed this. So the idea of `fit-content` was to represent the sizing formula of `auto` floats and tables so that it can be applied to other things, like regular blocks. And `fit-content(100%)` and `fit-content` are conceptually the same thing, assuming zero MBP; it would make sense if their behavior were essentially interchangeable. So the behavior here might need clarification in the spec, but unless there's some compelling reason not to, they should behave like `auto` does when it is interpreted as shrinkwrapping. Which is to say, contribute as for `max-content` when calculating max-content intrinsic sizes, and contribute as for `min-content` when contributing min-content sizes.

**Comment by fantasai** [comment]
Proposed edits:

```
diff --git a/css-sizing-3/Overview.bs b/css-sizing-3/Overview.bs
index b609ab2..325320b 100644
--- a/css-sizing-3/Overview.bs
+++ b/css-sizing-3/Overview.bs
@@ -444,6 +444,11 @@ Sizing Values: the <<length-percentage>>, ''width/auto'' | ''max-width/none'', '
                        i.e.
                        <code>min(''width/max-content'', max(''width/min-content'', <<length-percentage>>))</code>;
                        otherwise behaves as the property’s <a>initial value</a>.
+                       For the purpose of calculating <a>min-content contributions</a>,
+                       behaves as ''min-content'';
+                       for <a>max-content contributions</a>,
+                       as ''max-content''--
+                       clamped, if the argument is a <<length>>, by that length.
 
                        Negative values are invalid.
        </dl>
```

[css-sizing-4 is an experimental mess, hence the DO NOT IMPLEMENT warning; working against css-sizing-3]

**Comment by fantasai** [comment]
Also, I'm wondering if these values shouldn't be restricted to width/height, rather than also being valid on min/max-width/height? I'll file a separate issue on that...

**Comment by fantasai** [comment]
Oops, nevermind. We already resolved on that in https://www.w3.org/mid/CADhPm3sCDa0Uj3p4EE6GyoX-WBNJSzKn3tEOm5CWWPOCkGMq9A@mail.gmail.com :)

**Comment by dbaron** [comment]
I don't think that edit's handling of values that have no `<percentage>` part is correct -- for those, you just want to compute the single resulting length from the formula and use that for both the min-content contribution and the max-content contribution.  (For both contributions, this value might be the min-content size, it might be the length, and it might be the max-content size.)

I'm still unsure about whether it's really a good solution for values that do have a `<percentage>` part, although it is simple.

**Comment by fantasai** [comment]
Um, good point. :) Updated text:

```
			For <a>intrinsic size contributions</a>,
			''fit-content(<<length>>)'' behaves as the resulting length;
			''fit-content(<<percentage>>)'' behaves
			as ''width/min-content'' for <a>min-content contributions</a>,
			and as ''width/max-content'' for <a>max-content contributions</a>.
```

> I'm still unsure about whether it's really a good solution for values that do have a `<percentage>` part, although it is simple.

I'm open to suggestions, but in the absence of a clear benefit for another behavior, I think consistency with `auto` is a good principle to follow. So imho we'd need a) a solid proposal and b) use cases to justify any divergence. I don't have either atm...

**Comment by dbaron** [comment]
OK, I guess, except that that needs slight rewording to account for `calc()` expressions that are combinations of `<length>` and `<percentage>` -- wording such as replacing "`fit-content(<percentage>)`" with "other cases (values with percentages)".

**Comment by fantasai** [comment]
@dbaron reviews are the best reviews. :) Tentatively checked in.

**Comment by Loirooriol** [comment]
Why not rely on https://drafts.csswg.org/css-sizing-3/#percentage-sizing ?

Like if you have `min-width: fit-content(calc(100px + 10%))`, resolve the percentage against zero and treat it as `fit-content(100px)`.

For `max-width: fit-content(calc(100px + 10%))` and non-replaced `width: fit-content(calc(100px + 10%))`, treat them as `initial`.

If I have an element with `width: 1000px` and an intrinsic size of 10px, I think it wouldn't be consistent to say that adding `max-width: min(max-content, max(min-content, 0%))` results in an max-content contribution of `1000px` according to [§5.1.a](https://drafts.csswg.org/css-sizing-3/#non-replaced-percentage-contribution), but that `max-width: fit-content(0%)` results in an intrinsic contribution of `10px` according to fantasai's proposal.

IMO `fit-content(arg)` should be a perfect synonym of `min(max-content, max(min-content, arg))`, though maybe this could be achieved by treating `min`/`max` differently than `calc` in https://drafts.csswg.org/css-sizing-3/#percentage-sizing

**Comment by tabatkins** [comment]
We simplified to just rely on the behavior of the plain argument, as Oriol suggested. That is, `width: fit-content(50%)` behaves the same as `width: 50%;` in all circumstances, just clamped by min/max-content.

Agenda+ to confirm that this resolves dbaron's and Oriol's concerns.

**Comment by dbaron** [comment]
Where did the text that defines this actually end up?  I'm having trouble finding it.

**Comment by astearns** [comment]
The CSS Working Group just discussed Inline-axis intrinsic sizing for fit-content(%).
The full IRC log of that discussion
<fantasai> Topic: Inline-axis intrinsic sizing for fit-content(%)
<fantasai> github: https://github.com//issues/5119
<fantasai> https://github.com//issues/3731#issuecomment-661408877
<emilio> astearns: so the edits are in the spec already, is it just confirming that it's good?
<emilio> fantasai: yeah, oriol suggested to handle the fit-content argument like other percentages
<emilio> oriol: fit-content() is just syntax sugar for nested min/max functions with min-content / max-content on the values
<emilio> ... so they should behave the same as the expanded form
<fantasai> https://github.com/w3c/csswg-drafts/commit/6622d44241d41b38d19eed89fa796826f2121dcb
<emilio> dbaron: should probably review offline, though should take me about 10 mins
<emilio> astearns: let's do that

**Comment by astearns** [comment]
@dbaron I think it's the commit above? https://github.com/w3c/csswg-drafts/commit/6622d44241d41b38d19eed89fa796826f2121dcb

**Comment by dbaron** [comment]
That commit appears to *remove* the definition of how intrinsic sizes are handled.  It's not clear to me what definition then takes over -- I couldn't find anything relevant in the spec.

**Comment by fantasai** [comment]
@dbaron 

>  where the <length-percentage> argument is resolved exactly as for <length-percentage> values standing alone

Ends up referring to whatever the relevant layout module says about lengths and percentages, but also  https://drafts.csswg.org/css-sizing-3/#intrinsic
